Add Parcel Layer

rajadain opened the pull request:
## Overview

Adds Parcel Layer via the NC One Map ArcGIS API.

Closes #33 

### Demo

~~https://user-images.githubusercontent.com/1430060/187294794-54343556-8989-4515-a6cd-8ca3f50b9638.mp4~~ Old demo. Replaced with:

https://user-images.githubusercontent.com/1430060/187306338-8a1ddc78-346d-4c60-aaea-c27fba97947b.mp4

### Notes

~~The performance, despite all the tweaks, is quite dismal.~~ Improved in [952496a](https://github.com/azavea/iow-boundary-tool/pull/50/commits/952496a62d1f03abc216aa83d1059cc38718c5cb) and [36d5585](https://github.com/azavea/iow-boundary-tool/pull/50/commits/36d558543b59d34838a95c8792f38c5295c348e5)

~~One avenue to investigate may be to fetch images from the [MapServer](https://services.nconemap.gov/secure/rest/services/NC1Map_Parcels/MapServer/1) rather than features from the [FeatureServer](https://services.nconemap.gov/secure/rest/services/NC1Map_Parcels/FeatureServer/1).~~ Implemented in [952496a](https://github.com/azavea/iow-boundary-tool/pull/50/commits/952496a62d1f03abc216aa83d1059cc38718c5cb).

~~Another improvement we can make is to show the Parcel Centroids, which are also available as a layer, at lower zoom levels. This will give users _something_ rather than nothing.~~ Implemented in [1c7c41c](https://github.com/azavea/iow-boundary-tool/pull/50/commits/1c7c41c69da518e7a7b77f534d0c69e70c25f005).

## Testing Instructions

- Check out this branch and `server`
- Go to http://localhost:4545/draw
  - [ ] Ensure you see a disabled Parcel data layer button in the sidebar
- Zoom in to the map
  - [ ] Ensure the button turns on once you are zoomed in enough
  - [ ] Ensure the layer appears, as styled in Figma, when you are zoomed in enough
  - [ ] Ensure you can toggle the layer on and off while zoomed in enough

## Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [x] `README.md` updated if necessary to reflect the changes
- [x] CI passes after rebase

rajadain commented:
Switched to image based layer in [952496a](https://github.com/azavea/iow-boundary-tool/pull/50/commits/952496a62d1f03abc216aa83d1059cc38718c5cb), which is markedly faster. Now exploring if we can change the styling from the default hot pink to our preferred yellow 600:

![image](https://user-images.githubusercontent.com/1430060/187301130-79aff272-8463-4d68-ac02-c75617e1b88f.png)

rajadain commented:
Added Yellow 600 / Gold styling in [36d5585](https://github.com/azavea/iow-boundary-tool/pull/50/commits/36d558543b59d34838a95c8792f38c5295c348e5)

![image](https://user-images.githubusercontent.com/1430060/187304422-b9f499fe-d52a-46e0-b287-1fc990604153.png)

It makes the requests a tad slower, but still tolerable I think.

rajadain commented:
Added the ability to see parcel centroids on lower zoom levels in [1c7c41c](https://github.com/azavea/iow-boundary-tool/pull/50/commits/1c7c41c69da518e7a7b77f534d0c69e70c25f005) :

![image](https://user-images.githubusercontent.com/1430060/187305759-41c7fca4-0327-4f28-8a0b-474291fbadb5.png)

Not sure how useful this is, or if it should be styled differently.

rajadain commented:
New demo video that shows everything:

https://user-images.githubusercontent.com/1430060/187306338-8a1ddc78-346d-4c60-aaea-c27fba97947b.mp4

rajadain commented on src/app/src/components/Layers/ParcelLayer.js:
```diff
@@ -0,0 +1,69 @@
+                symbol: {
+                    type: 'esriSMS',
+                    style: 'esriSMSDiamond',
+                    color: [183, 121, 31, 255], // var(--chakra-colors-yellow-600)
```
It would be great to get this from Chakra UI somehow, like via `useTheme` or something, but I couldn't figure out a way to do that, so just added a comment for reference.

rajadain commented on src/app/src/components/Layers/ParcelLayer.js:
```diff
@@ -0,0 +1,69 @@
+            renderer: {
+                type: 'simple',
+                symbol: {
+                    type: 'esriSMS',
```
via https://developers.arcgis.com/web-map-specification/objects/esriSMS_symbol/

rajadain commented on src/app/src/components/Layers/ParcelLayer.js:
```diff
@@ -0,0 +1,69 @@
+                    style: 'esriSFSSolid',
+                    color: [0, 0, 0, 0], // No fill
+                    outline: {
+                        type: 'esriSLS',
```
via https://developers.arcgis.com/web-map-specification/objects/esriSLS_symbol/

rajadain commented on src/app/src/components/Layers/ParcelLayer.js:
```diff
@@ -0,0 +1,69 @@
+        new DynamicMapLayer({
+            url: PARCELS_LAYER_URL,
+            f: 'image',
+            layers: [0, 1],
```
Layer 0 is the Parcel Centroids, Layer 1 is the Parcel Boundaries

caseycesari replied:
Curios for @gatesgodin input, but I feel like the parcel centroids layer won't be useful for this application.

gatesgodin replied:
@caseycesari @rajadain I agree, parcel centroids are not very useful in this use case. Showing the parcel boundaries at the city or town level should be adequate. Colours and styling looks good at the parcel level.

rajadain commented on src/app/src/components/Sidebar.js:
```diff
@@ -83,14 +83,17 @@ function BasemapLayers() {
+                            visible={layers.includes(layer)}
+                            onChange={() => toggleVisibility(layer)}
+                            label={label}
+                            disabled={!!minZoom && mapZoom < minZoom}
```
Only disable if minZoom is defined and is greater than current mapZoom

rajadain reviewed (commented):
Left some comments assisting code review.

rajadain commented:
Tagging @gatesgodin for style feedback on the above.

caseycesari commented:
@rajadain Nice job figuring out a way to improve the performance! I'm glad this worked out because I wasn't looking forward to maintaining our own parcel layer.

mstone121 commented on src/app/src/components/Layers/ParcelLayer.js:
```diff
@@ -0,0 +1,69 @@
+    },
+];
+
+export default function ParcelLayer() {
+    const layers = useSelector(state => state.map.layers);
+    const showLayer = useMemo(() => layers.includes('PARCELS'), [layers]);
+
+    return showLayer ? <RenderEsriLayer /> : null;
+}
```
I created a hook called `useMapVisibility` in https://github.com/azavea/iow-boundary-tool/pull/46 that encapsulates some of this functionality, but perhaps it makes more sense to introduce it here since I'm no longer using it in #46.

```js
export function useLayerVisibility(layer) {
    return useSelector(state => state.map.layers).includes(layer);
}
```

rajadain replied:
Nice! Should this be wrapped in `useMemo` as the original is?

rajadain replied:
like:

```js
export function useLayerVisibility(layer) {
    const layers = useSelector(state => state.map.layers)
    const showLayer = useMemo(() => layers.includes('PARCELS'), [layers]);

    return showLayer;
}
```

or does `useMemo` only make sense in the context of a React Component?

mstone121 replied:
I'm not sure. Since the returned value is a boolean which is a primitive type, I think it's already memoized (i.e. it won't cause re-render when passed as a prop and it's value hasn't chaged, but it has been recalculated). 

I figured we could remove the hook and add it back if there are noticable performance issues.

rajadain replied:
Naïve implementation (without `useMemo`) in 44bbd4b

mstone121 reviewed (approved):
Nice work! I left a couple of comments, nothing major.

rajadain commented on src/app/src/components/DrawMap/useTrackMapZoom.js:
```diff
@@ -0,0 +1,14 @@
+export default function useTrackMapZoom() {
+    const dispatch = useDispatch();
+
+    const mapEvents = useMapEvents({
```
Fixed in e0a9e9e

rajadain commented:
Thanks for all the reviews! Squashed fixups and updated the CHANGELOG. I added a revert commit for the Parcel Centroids work instead of completely removing it, so it'll be preserved in git history for reference. Merging this now.
